sugondat-chain/runtimes/sugondat-kusama/tests/integration_test.rs

```diff
@@ -0,0 +1,137 @@
+use frame_support::traits::tokens::Precision;
+use sp_block_builder::runtime_decl_for_block_builder::BlockBuilderV6;
+use sp_runtime::{
+    generic::SignedPayload,
+    transaction_validity::{
+        InvalidTransaction, TransactionSource, TransactionValidity, TransactionValidityError,
+    },
+    BuildStorage, MultiSignature,
+};
+
+use codec::Encode;
+use frame_support::traits::fungible::Balanced;
+use sp_core::{crypto::Pair, sr25519};
+use sp_transaction_pool::runtime_api::runtime_decl_for_tagged_transaction_queue::TaggedTransactionQueueV3;
+use sugondat_kusama_runtime::{
+    Address, Hash, MaxBlobSize, MaxBlobs, MaxTotalBlobSize, Runtime, RuntimeCall, SignedExtra,
+    UncheckedExtrinsic,
+};
+
+use sugondat_kusama_runtime::*;
+
+pub fn new_test_ext() -> sp_io::TestExternalities {
+    frame_system::GenesisConfig::<Runtime>::default()
+        .build_storage()
+        .unwrap()
+        .into()
+}
+
+// Needed to be called inside Externalities
+pub fn submit_blob_utx(namespace_id: u32, blob: Vec<u8>) -> UncheckedExtrinsic {
```

Comment:
Ok, I spoke too fast. Let's rewind:

1. By the name of the function, I first assumed that it submits a blob.
2. Then reading the code I noticed the return type and the `new_signed` at the end of the function, and then I assumed that the function is pure and just returns the extrinsic.
3. then I read the code closer and realized that it's not so simple, because the function does create an extrinsic but it also modifies some global state.

Generally, it's a good practice to split up modification of state and "pure" functions that just return some data based on it's inputs.

Here is how you can classify code from the least dangerous to most:

1. code that doesn't read any state (returns a constant)
1. code that returns something based on passed arguments ("pure")
1. code that performs modifications to something passed via arguments
1. code that returns something based on the thread/task local state
1. code that performs modifications to thread/task local state
1. code that returns something based on the global state
1. code that performs modifications to global state

The "dangerousness" is correlated with power of the code (e.g. number one is harmless) and the ease of reasoning of the code (it's generally harder to fit all possible states the program can be in for the code at the bottom of the least relative to the code at the top of the list).

Here, we are at number 6 (sometimes it's hard to tell the difference between thread and global state. I am luckily familiar with the underlying impl) here.

Ideally code that are at the bottom of the list, should scream about it's "dangerousness". This "screaming" can take different forms:

- It can be naming and comments. `prepare_environment_and_mk_submit_blob_utx`.
- Isolating the code that performs into it's own function. (This one may be tricky, because, you need to make sure that the code in question is obvious and extremely visible and sometimes wrapping it in a function may hide it. So you your judgement).
- One advantage of Rust, is that it makes you to specify the stuff that you modify and thus 3 may be identified quickly via the signature of the function having a &mut. Another is that it's coding culture makes you scrutinize things like `RefCell`.

Reply:
Thank you so much for the explanation! I really liked how you divided functions by their level of danger. Yes, that particular function is indeed a complete mess in that regard! I will try to update it now, hopefully, making it less dangerous, and if I can't, at least making it more explicit!